### PR TITLE
Cache JS inferred class type symbol

### LINKED ISSUE #32997: JS Crash: global type reference to constructor function
From chrome-devtools-frontend's failure in https://github.com/microsoft/TypeScript/pull/32991:

```js
// @Filename: test.js
var SDK = {}
SDK.SourceMap = function () { }
SDK.SourceMap.prototype = { compiledURL() {} }

// @Filename: other.js
/**
 * @param {SDK.SourceMap} sourceMap
 */
function no(sourceMap) {
  return sourceMap.wat
}
```

**Expected behavior:**
no crash and a bunch of errors

**Actual behavior:**
assert from merging a symbol into itself. it looks like both the `function () {}` and `SDK.SourceMap` lhs of assignment get the prototype method in the binder.

### PR DESCRIPTION
**Pull request by sandersn**
Note that many sources merge into a single target, so the *source* [links] is the one that caches the merged target.

The reason this is a problem is not that many sources merge into a single target, but that both getTypeOfSymbol and getDeclaredTypeOfSymbol
end up calling mergeJSSymbols with the same [source,target] pair. The merge should not happen twice.

Fixes #32997

### CODE REVIEW

File: src/compiler/types.ts [hunk 1]
```diff
@@ -3770,30 +3770,31 @@ namespace ts {
...
+        immediateTarget?: Symbol;                 // Immediate target of an alias. May be another alias. Do not access directly, use `checker.getImmediateAliasedSymbol` instead.
+        target?: Symbol;                          // Resolved (non-alias) target of an alias
+        type?: Type;                              // Type of value symbol
+        uniqueESSymbolType?: Type;                // UniqueESSymbol type for a symbol
+        declaredType?: Type;                      // Type of class, interface, enum, type alias, or type parameter
+        resolvedJSDocType?: Type;                 // Resolved type of a JSDoc type reference
+        typeParameters?: TypeParameter[];         // Type parameters of type alias (undefined if non-generic)
+        outerTypeParameters?: TypeParameter[];    // Outer type parameters of anonymous object type
+        instantiations?: Map<Type>;               // Instantiations of generic type alias (undefined if non-generic)
+        inferredClassSymbol?: TransientSymbol;    // Symbol of an inferred ES5 constructor function
```
Comment (sandersn) [inline src/compiler/types.ts]: ignore whitespace to highlight the actual change
(I added some whitespace on the other lines)

**Comment by sandersn** [comment]
Uggg breaks chainedPrototypeAssignment.ts

```js
A.prototype = B.prototype = { ... }
```

I'm not convinced this is sane code, though. Accepting the baselines for now while I think about it.

**Comment by sandersn** [comment]
I thought about it. It's not real code, just a test case I came up with, so I don't feel bad about breaking it.

**Comment by sandersn** [comment]
Adding @andrewbranch and @orta in case they want to think about caching in the compiler. =)

**Comment by weswigham** [comment]
> I'm not convinced this is sane code, though. Accepting the baselines for now while I think about it.

I think the new behavior is arguably more sane - references to both `A` and `B` now both reflect the same symbol.

Actually, wait.... they're still different constructors, just with separate prototypes, which means... ick. If `A` and `B` have differing signatures, this is going to wholly substitute `A` with `B` - ick.

@sandersn I think the `inferredClassType` cache may need to be a map with the appropriate `target` symbol id, so the same `source` can be merged into multiple `target`s.

**Comment by sandersn** [comment]
@weswigham mind taking another look?

File: tests/baselines/reference/chainedPrototypeAssignment.types [hunk 2]
```diff
@@ -14,8 +14,8 @@ var a = new mod.A()
 >A : typeof A
 
 var b = new mod.B()
->b : A
->new mod.B() : A
+>b : B
+>new mod.B() : B
```
Comment (andrewbranch) [inline tests/baselines/reference/chainedPrototypeAssignment.types]: This looks better 😄

**Comment by sandersn** [comment]
Hooray! My insane test case continues to work~~ :alien: :squid: :crab:

**Pull request merged**